lxd/storage/utils.go

```diff
@@ -360,3 +384,253 @@ func GetStorageResource(path string) (*api.ResourcesStoragePool, error) {
 
 	return &res, nil
 }
+
+// VolumeTypeNameToType converts a volume type string to internal code.
+func VolumeTypeNameToType(volumeTypeName string) (int, error) {
+	switch volumeTypeName {
+	case db.StoragePoolVolumeTypeNameContainer:
+		return db.StoragePoolVolumeTypeContainer, nil
+	case db.StoragePoolVolumeTypeNameImage:
+		return db.StoragePoolVolumeTypeImage, nil
+	case db.StoragePoolVolumeTypeNameCustom:
+		return db.StoragePoolVolumeTypeCustom, nil
+	}
+
+	return -1, fmt.Errorf("invalid storage volume type name")
+}
+
+// VolumeDBCreate creates a volume in the database.
+func VolumeDBCreate(s *state.State, poolName string, volumeName, volumeDescription string, volumeTypeName string, snapshot bool, volumeConfig map[string]string) error {
+	// Convert the volume type name to our internal integer representation.
+	volumeType, err := VolumeTypeNameToType(volumeTypeName)
+	if err != nil {
+		return err
+	}
+
+	// Load storage pool the volume will be attached to.
+	poolID, poolStruct, err := s.Cluster.StoragePoolGet(poolName)
+	if err != nil {
+		return err
+	}
+
+	// Check that a storage volume of the same storage volume type does not
+	// already exist.
+	volumeID, _ := s.Cluster.StoragePoolNodeVolumeGetTypeID(volumeName, volumeType, poolID)
+	if volumeID > 0 {
+		return fmt.Errorf("a storage volume of type %s does already exist", volumeTypeName)
+	}
+
+	// Make sure that we don't pass a nil to the next function.
+	if volumeConfig == nil {
+		volumeConfig = map[string]string{}
+	}
+
+	// Validate the requested storage volume configuration.
+	err = VolumeValidateConfig(poolName, volumeConfig, poolStruct)
+	if err != nil {
+		return err
+	}
+
+	err = VolumeFillDefault(poolName, volumeConfig, poolStruct)
+	if err != nil {
+		return err
+	}
+
+	// Create the database entry for the storage volume.
+	_, err = s.Cluster.StoragePoolVolumeCreate("default", volumeName, volumeDescription, volumeType, snapshot, poolID, volumeConfig)
+	if err != nil {
+		return fmt.Errorf("Error inserting %s of type %s into database: %s", poolName, volumeTypeName, err)
+	}
+
+	return nil
+}
+
+// SupportedPoolTypes the types of pools supported.
+var SupportedPoolTypes = []string{"btrfs", "ceph", "cephfs", "dir", "lvm", "zfs"}
```

Comment:
We'll want to make sure we don't rely on this for any of the new code as we won't want such a list in the main `storage` package. Only `drivers` should tell us what backend exists.

Reply:
We'll actually want some kind of `SupportedDrivers()` function the drivers package which should give us:
 - Name of the driver
 - Version
 - If it's currently usable (tools/kernel support present)
 - What can it store (list of volume types)
 - Whether it's a local storage or remote

This is all information we need for `lxd init` so we can offer the right set of storage drivers to the user at setup time. We may extend that description struct with some more bits based on what's needed to driver the `lxd init` questions.